Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp

```diff
@@ -0,0 +1,349 @@
+        loadRegionalLocalization(nullptr, nullptr, nullptr, emptyString(), emptyString());
+        return;
+    }
+
+    RELEASE_LOG_DEBUG(Extensions, "Loaded default locale %s", defaultLocaleString.utf8().data());
+
+    auto bestLocaleString = webExtension.bestMatchLocale();
+    auto defaultLocaleComponents = parseLocale(defaultLocaleString);
+    auto bestLocaleComponents = parseLocale(bestLocaleString);
+    auto bestLocaleLanguageOnlyString = bestLocaleComponents.languageCode;
+
+    RELEASE_LOG_DEBUG(Extensions, "Best locale is %s", bestLocaleString.utf8().data());
+
+    RefPtr<JSON::Object> languageJSON;
+    if (!bestLocaleLanguageOnlyString.isEmpty() && bestLocaleLanguageOnlyString != defaultLocaleString) {
+        languageJSON = localizationJSONForWebExtension(webExtension, bestLocaleLanguageOnlyString);
+        if (languageJSON)
+            RELEASE_LOG_DEBUG(Extensions, "Loaded language-only locale %s", bestLocaleLanguageOnlyString.utf8().data());
+    }
+
+    RefPtr<JSON::Object> regionalJSON;
+    if (bestLocaleString != bestLocaleLanguageOnlyString && bestLocaleString != defaultLocaleString) {
+        regionalJSON = localizationJSONForWebExtension(webExtension, bestLocaleString);
+        if (regionalJSON)
+            RELEASE_LOG_DEBUG(Extensions, "Loaded regional locale %s", bestLocaleString.utf8().data());
+    }
+
+    loadRegionalLocalization(regionalJSON, languageJSON, defaultLocaleJSON, bestLocaleString);
+}
+
+WebExtensionLocalization::WebExtensionLocalization(RefPtr<JSON::Object> localizedJSON, const String& uniqueIdentifier)
+{
+    ASSERT(!uniqueIdentifier.isEmpty());
+    ASSERT(localizedJSON);
+
+    auto localeString = emptyString();
+    if (RefPtr predefinedMessages = localizedJSON->getObject(predefinedMessageUILocale))
+        localeString = predefinedMessages->getString(messageKey);
+
+    loadRegionalLocalization(localizedJSON, nullptr, nullptr, localeString, uniqueIdentifier);
+}
+
+WebExtensionLocalization::WebExtensionLocalization(RefPtr<JSON::Object> regionalLocalization, RefPtr<JSON::Object> languageLocalization, RefPtr<JSON::Object> defaultLocalization, const String& withBestLocale, const String& uniqueIdentifier)
+{
+    loadRegionalLocalization(regionalLocalization, languageLocalization, defaultLocalization, withBestLocale, uniqueIdentifier);
+}
+
+void WebExtensionLocalization::loadRegionalLocalization(RefPtr<JSON::Object> regionalLocalization, RefPtr<JSON::Object> languageLocalization, RefPtr<JSON::Object> defaultLocalization, const String& withBestLocale, const String& uniqueIdentifier)
+{
+    m_locale = WebCore::Locale::create(AtomString { withBestLocale });
+    m_localeString = withBestLocale;
+    m_uniqueIdentifier = uniqueIdentifier;
+
+    RefPtr<JSON::Object> localizationJSON = predefinedMessages(defaultLocalization ?: regionalLocalization ?: languageLocalization);
+    localizationJSON = JSONWithLowercaseKeys(localizationJSON);
```

Comment:
This is not needed since it is only predefined messages for far, and those are already lowercase.
```suggestion
```